OSDOCS-2932: OSDK-1760: digest bundle for disconnected environments (#42315)
openshift/openshift-docs: modules/olm-enabling-operator-restricted-network.adoc

// TODO: Include more info w/ better steps on how to do this:
//* You must understand the {product-title} proxy configuration.

For the CSV requirements, you can make the following changes as the Operator author. The following procedure uses the Memcached Operator created in the Operator SDK tutorials for Go, Ansible, and Helm.

.Prerequisites

* An Operator project with a CSV.

.Procedure

. Replace hard-coded image references with environment variables in the following files:

> ** For Go- and Ansible-based Operator projects, update the `roles/memcached/tasks/main.yml` file:

Reviewer: Go-based Operator doesn't have the file roles/memcached/tasks/main.yml. Can update the image in example file [controllers/memcached_controller.go](https://docs.openshift.com/container-platform/4.9/operators/operator_sdk/golang/osdk-golang-tutorial.html#osdk-golang-implement-controller_osdk-golang-tutorial)

Update the image to be the digest type.
Image: "memcached:1.4.36-alpine",
Image:   "docker.io/memcached@sha256:00b68b00139155817a8b1d69d74865563def06b3af1e6fc79ac541a1b2f6b961",

+
[source,diff]
----
...
spec:
  containers:
  - name: memcached
    command:
    - memcached
    - -m=64
    - -o
    - modern
    - -v
-   image: "docker.io/memcached:1.4.36-alpine"
+   image: "{{ lookup('env', 'RELATED_IMAGE_MEMCACHED') }}"
    ports:
      - containerPort: 11211
...
----

** For Helm-based Operator projects, update the `helm-charts/memchached/values.yaml` file:
+
[source,diff]
----
  ## Memcached image and tag
  ## ref: https://hub.docker.com/r/library/memcached/tags/
  ##
- image: memcached:1.5.20
+ image: "{{ lookup('env', 'RELATED_IMAGE_MEMCACHED') }}"
...
----

.Give the images and tags used by the Operator by using an environment variable in the `config/manager/manager.yaml` file:
+
[source,yaml]
----
...
spec:
  ...
    spec:
      ...
      containers:
      - command:
        - /manager
        ...
        env:
        - name: RELATED_IMAGE_MEMCACHED
          value: "quay.io/example.com/image:0.0.1"
----

. Use the `make bundle` command and set `USE_IMAGE_DIGESTS` to `true` to automatically list any related images that your Operator might require and reference each image by a digest (SHA), not a tag:
+
[source,terminal]
----

> $ make bundle USE_IMAGE_DIGESTS=true

Reviewer: Should add one step before "make bundle USE_IMAGE_DIGESTS=true" because of bug [2047552](https://bugzilla.redhat.com/show_bug.cgi?id=2047552)

Update the Makefile
" $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)" 

to be 

"$(KUSTOMIZE) build config/manifests | operator-sdk generate bundle -q --overwrite --version $(VERSION) $(BUNDLE_GEN_FLAGS)"

Reviewer: Others LGTM

// TODO: Include more info w/ better steps on how to do this:
//* You must understand the {product-title} proxy configuration.

For the CSV requirements, you can make the following changes as the Operator author. The following procedure uses the Memcached Operator created in the Operator SDK tutorials for Go, Ansible, and Helm.

.Prerequisites

* An Operator project with a CSV.

.Procedure

> . Add the images and tags used by the Operator by setting an environment variable, such as `RELATED_IMAGE_MEMCACHED`, in the `config/manager/manager.yaml` file:

Reviewer 2: Specifically, I would call out that these are for additional images used by the operator. For example with Memcached operator, the operator image itself will be automagically resolved to a digest, but additional images, such as the actual Memcached image, will need to be specified here so they can be resolved.

+
.Example `config/manager/manager.yaml`
[%collapsible]
====
[source,yaml]
----
...
spec:
  ...
    spec:
      ...
      containers:
      - command:
        - /manager
        ...
        env:
        - name: RELATED_IMAGE_MEMCACHED

>           value: "docker.io/memcached@sha256:00b68b00139155817a8b1d69d74865563def06b3af1e6fc79ac541a1b2f6b961"

Reviewer 2: You do not need the digest here, just the normal tagged image. It will be resolved to a digest when you run `make bundle`.
```suggestion
          value: "memcached:1.4.36-alpine"
```

----
====

. Replace hard-coded image references with environment variables in the following files:

* For Go-based Operator projects, update the `controllers/memcached_controller.go` file with the following changes:
+
.`controllers/memcached_controller.go` diff
[%collapsible]
====
[source,diff]
----
...
// deploymentForMemcached returns a memcached Deployment object
...
	Spec: corev1.PodSpec{
        	Containers: []corev1.Container{{
-			Image:   "memcached:1.4.36-alpine",

> +			Image:   "{{ lookup('env', 'RELATED_IMAGE_MEMCACHED') }}",

Reviewer: Go type operator doesn't have the lookup function. Should update the Image to the digest type: 
Update the image to be the digest type.
Image: "memcached:1.4.36-alpine",
To be
Image: "docker.io/memcached@sha256:00b68b00139155817a8b1d69d74865563def06b3af1e6fc79ac541a1b2f6b961",

> +			Image:   "docker.io/memcached@sha256:00b68b00139155817a8b1d69d74865563def06b3af1e6fc79ac541a1b2f6b961",

Reviewer 2: This will pull the image with digest from the environment:

```suggestion
+			Image:   os.Getenv("RELATED_IMAGE_MEMCACHED"),
```

Optionally, you can mention that `os.Getenv` will return an empty string if the variable is unset, so users may want to validate that the env variable is present first.

			Ports: []corev1.ContainerPort{{
...
----
====

* For Ansible-based Operator projects, update the `roles/memcached/tasks/main.yml` file with the following changes:
+
.`roles/memcached/tasks/main.yml` diff
[%collapsible]
====
[source,diff]
----
...
spec:
  containers:
  - name: memcached
    command:
    - memcached
    - -m=64
    - -o
    - modern
    - -v
-   image: "docker.io/memcached:1.4.36-alpine"
+   image: "{{ lookup('env', 'RELATED_IMAGE_MEMCACHED') }}"
    ports:
      - containerPort: 11211
...
----
====

* For Helm-based Operator projects, update the `helm-charts/memchached/values.yaml` file with the following changes:
+
.`helm-charts/memchached/values.yaml` diff
[%collapsible]
====
[source,diff]
----
  ## Memcached image and tag
  ## ref: https://hub.docker.com/r/library/memcached/tags/
  ##
- image: memcached:1.5.20
+ image: "{{ lookup('env', 'RELATED_IMAGE_MEMCACHED') }}"
...
----
====

. Update the `Makefile` file with the following changes:
+
.`Makefile` diff
[source,diff]
----
...
- $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)

> + $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle -q --overwrite --version $(VERSION) $(BUNDLE_GEN_FLAGS)

Reviewer 2: ```suggestion
+ $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle $(BUNDLE_GEN_FLAGS)
$(BUNDLE_GEN_FLAGS)
```

You will also need to instruct users to add the definition of `BUNDLE_GEN_FLAGS` as well. In a newly generated `Makefile` this will be at the top with the other variable definitions, but just needs to be before this line:

```Makefile
BUNDLE_GEN_FLAGS ?= -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)

# USE_IMAGE_DIGESTS defines if images are resolved via tags or digests
# You can enable this value if you would like to use SHA Based Digests
# To enable set flag to true
USE_IMAGE_DIGESTS ?= false
ifeq ($(USE_IMAGE_DIGESTS), true)
	BUNDLE_GEN_FLAGS += --use-image-digests
endif
```

// TODO: Include more info w/ better steps on how to do this:
//* You must understand the {product-title} proxy configuration.

For the CSV requirements, you can make the following changes as the Operator author. The following procedure uses the Memcached Operator created in the Operator SDK tutorials for Go, Ansible, and Helm.

.Prerequisites

* An Operator project with a CSV.

.Procedure

. Set an environment variable, such as `RELATED_IMAGE_MEMCACHED`, for any additional images used by the Operator in the `config/manager/manager.yaml` file:
+
.Example `config/manager/manager.yaml`
[%collapsible]
====
[source,yaml]
----
...
spec:
  ...
    spec:
      ...
      containers:
      - command:
        - /manager
        ...
        env:
        - name: RELATED_IMAGE_MEMCACHED
          value: "docker.io/memcached:1.4.36-alpine"
----
====

> . Pull the related images with digests by replacing hard-coded image references with environment variables in the following files:

Reviewer 3: Here I got a little confused bc after I see in the following files I kind of expected a more straightforward list. It might be better to either cut number 2 off as a sentence "Pull the related images with digests by replacing hard-coded image references with environment variables" and then have the files listed in the bullets below like you have them or you could list out the files somehow.

> * For Go-based Operator projects, update the `controllers/memcached_controller.go` file with the following changes:

Reviewer 3: This kind of goes along with my comment above. I think maybe only one or the other should have "with the following changes" it just seems a bit confusing to me. Do users need to make the changes in each file?

It could be something as simple as adding a small addition to number 2 that discusses what type of operator project gets updated maybe? Just a suggestion:

> Pull the related images with digests by replacing hard-coded image references with environment variables for Go, Ansible, and Helm-based Operator Projects. 

And then i think you could leave the bullets as is.

Reviewer 4: It would be better to be more direct:
```suggestion
* For Go-based Operator projects, add the environment variable to the `controllers/memcached_controller.go` file as shown in the following example:
```

+

> .`controllers/memcached_controller.go` diff

Reviewer 4: ```suggestion
.Example `controllers/memcached_controller.go` file
```

Abbreviations won't translate well.

[%collapsible]
====
[source,diff]
----
...
// deploymentForMemcached returns a memcached Deployment object
...
	Spec: corev1.PodSpec{
        	Containers: []corev1.Container{{
-			Image:   "memcached:1.4.36-alpine",
+			Image:   os.Getenv("RELATED_IMAGE_MEMCACHED"),
			Name:    "memcached",
			Command: []string{"memcached", "-m=64", "-o", "modern", "-v"},
			Ports: []corev1.ContainerPort{{
...
----
+

> The `os.Getenv` function returns an empty string in the variable is not set. Set the environment variable before changing the file.

Reviewer 4: Where do you set this environment variable? I don't see it. Please be clearer what it is in the code snippet by enclosing it in angled brackets and add a callout that's explicit about what the user's supposed to do.

====

> * For Ansible-based Operator projects, update the `roles/memcached/tasks/main.yml` file with the following changes:

Reviewer 4: The same sort of comments apply to these other options.

+
.`roles/memcached/tasks/main.yml` diff
[%collapsible]
====
[source,diff]
----
...
spec:
  containers:
  - name: memcached
    command:
    - memcached
    - -m=64
    - -o
    - modern
    - -v
-   image: "docker.io/memcached:1.4.36-alpine"
+   image: "{{ lookup('env', 'RELATED_IMAGE_MEMCACHED') }}"
    ports:
      - containerPort: 11211
...
----
====

* For Helm-based Operator projects, update the `helm-charts/memchached/values.yaml` file with the following changes:
+
.`helm-charts/memchached/values.yaml` diff
[%collapsible]
====
[source,diff]
----
  ## Memcached image and tag
  ## ref: https://hub.docker.com/r/library/memcached/tags/
  ##
- image: memcached:1.5.20
+ image: "{{ lookup('env', 'RELATED_IMAGE_MEMCACHED') }}"
...
----
====

> . Add the `BUNDLE_GEN_FLAGS` variable definition to your `Makefile` file and make the following changes:

Reviewer 3: It seems like there is a lot of adding/editing in these yaml files. WDYT about adding in a line at like #1 or #2 or even dropping a line before the _operator requirements for supporting disconnected code_ that says in order to do this you will have to edit/add to yaml files.

Reviewer 3: Also, I would remove "file" and just say:

> Add the `BUNDLE_GEN_FLAGS` variable definition to your `Makefile` and make the following changes:

+
.`Makefile` diff
[source,diff]
----
...
BUNDLE_GEN_FLAGS ?= -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)

# USE_IMAGE_DIGESTS defines if images are resolved via tags or digests
# You can enable this value if you would like to use SHA Based Digests
# To enable set flag to true
USE_IMAGE_DIGESTS ?= false
ifeq ($(USE_IMAGE_DIGESTS), true)
	BUNDLE_GEN_FLAGS += --use-image-digests
endif

-  $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)
+  $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle $(BUNDLE_GEN_FLAGS)
...
----

> . Use the `make bundle` command and set `USE_IMAGE_DIGESTS` to `true` to automatically update your Operator image to use a digest (SHA) and not a tag:

Reviewer 4: ```suggestion
.  To automatically update your Operator image to use a digest (SHA) and not a tag, run the `make bundle` command to set `USE_IMAGE_DIGESTS` to `true`:
```

+
[source,terminal]
----
$ make bundle USE_IMAGE_DIGESTS=true
----
+
[TIP]
====

> Alternatively, you can update your Operator project manually by using SHA references to related images in two places in the CSV for your Operator:

Reviewer 4: This is an awfully long tip, and it's not rendering well. Is there a discrete use case for doing this? Is a stakeholder insisting that you add this? If there's not a compelling reason to add this info, you only need to show the way that you _want_ users to do a thing.

If there's a specific time when it's better to use this method, you should present it as an option, like you did with the file changes, and explain when to do which.

> Alternatively, you can update your Operator project manually by using SHA references to related images in two places in the CSV for your Operator:

Reviewer 4: ```suggestion
Alternatively, you can update your Operator project manually by using SHA references to related images in two places in the CSV <noun_goes_here> for your Operator:
```

> * Update `spec.relatedImages`:

Reviewer 3: Just a thought: I am wondering if it would be better for these to be numbered instead of bullets since you have "two places" in the initial sentence.

Reviewer 3: or if you want to leave the bullet, you could say something like "The first place you can update ..."

// TODO: Include more info w/ better steps on how to do this:
//* You must understand the {product-title} proxy configuration.

For the CSV requirements, you can make the following changes as the Operator author. The following procedure uses the Memcached Operator created in the Operator SDK tutorials for Go, Ansible, and Helm.

.Prerequisites

* An Operator project with a CSV.

.Procedure

. Add the `BUNDLE_GEN_FLAGS` variable definition to your `Makefile` file and make the following changes:
+
.`Makefile` diff
[source,diff]
----
...
-  $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)
+  $(KUSTOMIZE) build config/manifests | operator-sdk generate bundle $(BUNDLE_GEN_FLAGS)
+  $(BUNDLE_GEN_FLAGS)

BUNDLE_GEN_FLAGS ?= -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)

# USE_IMAGE_DIGESTS defines if images are resolved via tags or digests
# You can enable this value if you would like to use SHA Based Digests
# To enable set flag to true
USE_IMAGE_DIGESTS ?= false
ifeq ($(USE_IMAGE_DIGESTS), true)
	BUNDLE_GEN_FLAGS += --use-image-digests

> endif

Author: @ryantking Is this correct?

Author: Also, do we need the second step where the `make bundle USE_IMAGE_DIGESTS=true`? Couldn't we just set `USE_IMAGE_DIGESTS ?= false` to `true` in line 45?

Reviewer: The "make bundle USE_IMAGE_DIGESTS=true" step should be the last step.

<1> Inject the images referenced by the Operator by using environment variables.
<2> Specify each image by a digest (SHA), not by an image tag.
<3> Also reference the Operator container image by a digest (SHA), not by an image tag.
<4> The `periodSeconds` value must be higher than the `timeoutSeconds` value when configuring probes. If left unspecified, the `timeoutSeconds` default value is `1`. The `periodSeconds` default value is `10`.

--
=====
====

. Set an environment variable, such as `RELATED_IMAGE_MEMCACHED`, for the additional images used by the Operator in the `config/manager/manager.yaml` file:
+
.Example `config/manager/manager.yaml`
[%collapsible]
====
[source,yaml]
----
...
spec:
  ...
    spec:
      ...
      containers:
      - command:
        - /manager
        ...
        env:
        - name: RELATED_IMAGE_MEMCACHED

>           value: "docker.io/memcached:1.4.36-alpine"

Author: @ryantking If we call `make bundle` in step 2, should this be the tag or the digest?

Reviewer: The "make bundle USE_IMAGE_DIGESTS=true" step should be the last step.
